runtime/lua/vim/func/_memoize.lua

```diff
@@ -33,31 +35,67 @@ local function resolve_hash(hash)
   return hash
 end
 
+--- Memoizes a function {fn} using {hash} to hash the arguments.
+---
+--- Internally uses a |lua-weaktable| to cache the results of {fn} meaning the
+--- cache will be invalidated whenever Lua does garbage collection.
+---
+--- The memoized function returns shared references so be wary about
+--- mutating return values.
+---
 --- @generic F: function
---- @param hash integer|string|fun(...): any
---- @param fn F
---- @param strong? boolean
---- @return F
-return function(hash, fn, strong)
+--- @param hash integer|string|function Hash function to create a hash to use as a key to
+---     store results. Possible values:
+---     - When integer, refers to the index of an argument of {fn} to hash.
+---     This argument can have any type.
+---     - When function, is evaluated using the same arguments passed to {fn}.
+---     - When `concat`, the hash is determined by string concatenating all the
+---     arguments passed to {fn}.
+---     - When `concat-n`, the hash is determined by string concatenating the
+---     first n arguments passed to {fn}.
+---
+--- @param fn F Function to memoize.
+--- @param strong? boolean Do not use a weak table
+--- @return F # Memoized version of {fn}
+--- @nodoc
+M.init = function(hash, fn, strong)
```

Comment:
Not sure if it's absolutely needed since the module is private, but I think we can prefix the methods in this module with an underscore to mark them as private (and that way you can remove the `@nodoc` annotation).